### PR TITLE
[HW] Introduce the `hw.instance_choice` op

### PR DESCRIPTION
**Pull request by nandor**
This PR introduces an op to represent instance alternatives in the HW dialect.

It brings in the definition and the syntax of the op. Subsequent changes will implement a FIRRTL source op and adjust the pipeline to handle then.

### CODE REVIEW

File: include/circt/Dialect/HW/HWStructure.td [hunk 1]
```diff
@@ -518,6 +518,40 @@ def InstanceOp : HWInstanceOpBase<"instance", [HWInstanceLike]> {
...
+    For the purposes of analyses and transformations, it is assumed that any of
+    the targets is a possibility.
+
+    Example:
+    ```mlir
+    %b = hw.instance_choice "inst" sym
+        @TargetDefault or
+        @TargetA if "A" or
+        @TargetB if "B"
+        (a: %a: i32) -> (b: i32)
```
Comment (rwy7) [inline include/circt/Dialect/HW/HWStructure.td:538]: not sure, but maybe you're missing a colon here
```suggestion
        : (a: %a: i32) -> (b: i32)
```

Comment (rwy7) [reply include/circt/Dialect/HW/HWStructure.td]: Oh apologies, I misread this as a trailing type, but it is the argument list, so no colon here.

**Review by rwy7 — APPROVED** [review]

**Review by uenoku — APPROVED** [review]
If we allow instanc_choise op to have an inner sym, how does XMRRef get lowered?

**Comment by nandor** [comment]
The plan is to ban properties/probes on these instance for now. The choices are likely to be simple leaf nodes (for example, instantiate an intrinsic instead of a clock gate black box), so that complexity is not immediately required. If we do decide to make cuts at higher levels, that will require serious extensions to XMRs/probes/properties.

**Pull request merged**